src/ActionBuilder/Cart/SetCustomShippingMethod.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+namespace BestIt\CommercetoolsODM\ActionBuilder\Cart;
+
+use BestIt\CommercetoolsODM\Mapping\ClassMetadataInterface;
+use Commercetools\Core\Model\Cart\Cart;
+use Commercetools\Core\Model\ShippingMethod\ShippingRate;
+use Commercetools\Core\Model\TaxCategory\TaxCategoryReference;
+use Commercetools\Core\Request\AbstractAction;
+use Commercetools\Core\Request\Carts\Command\CartSetCustomShippingMethodAction;
+
+/**
+ * Set custom shipping method
+ *
+ * @author Michel Chowanski <chowanski@bestit-online.de>
+ * @package BestIt\CommercetoolsODM\ActionBuilder\Cart
+ */
+class SetCustomShippingMethod extends CartActionBuilder
+{
+    /**
+     * Key for setting a custom shipping method
+     */
+    const IS_CUSTOM = 'IsCustomShippingMethod';
+
+    /**
+     * A PCRE to match the hierarchical field path without delimiter.
+     * @var string
+     */
+    protected $complexFieldFilter = 'shippingInfo/(.*)$';
+
+    /**
+     * Creates the update action for the given class and data.
+     *
+     * @param mixed $changedValue
+     * @param ClassMetadataInterface $metadata
+     * @param array $changedData
+     * @param array $oldData
+     * @param Cart $sourceObject
+     * @param string $subFieldName If you work on attributes etc. this is the name of the specific attribute.
+     *
+     * @return AbstractAction[]
+     */
+    public function createUpdateActions(
+        $changedValue,
+        ClassMetadataInterface $metadata,
+        array $changedData,
+        array $oldData,
+        $sourceObject,
+        string $subFieldName = ''
+    ): array {
+        $actions = [];
+
+        if (($state = $changedData['shippingMethodState'] ?? null) === static::IS_CUSTOM) {
+
+            $origin = $oldData['shippingInfo'];
+            $info = [
+                'shippingMethodName' => $changedData['shippingMethodName'] ?? $origin['shippingMethodName'],
+                'shippingRate' => array_replace_recursive($origin['shippingRate'], $changedData['shippingRate'] ?? []),
+                'taxCategory' => array_replace_recursive($origin['taxCategory'], $changedData['taxCategory'] ?? []),
+            ];
```

Comment:
Ich muss meine neuen Daten zum Teil auch mit den alten mergen. Grund: Einige Daten sind nicht wirklich geändert, wird aber für die Action benötigt. Einfaches Beispiel: 'currency'.

Beim Wechsel der Versandmethode bleibt die Währung i.d.R. die gleiche und kommt daher in `$changeData` nicht vor - wird aber zwingend für die Action benötigt. Hat sich diese nicht geändert, fügen wir diese anhand der alten Daten hinzu.

Reply:
Können wir nicht besser das ODM komplett anpassen, sodass die alten Daten mitgegeben werden?

Ich hatte das Problem auch schon mal.

Reply:
Verstehe ich nicht ganz. Die alten Daten werden ja mitgegeben - dadurch kann ich ja den "Merge" Prozess durchführen. In der Regel braucht eine Action nur die geänderten Werte - hier ist mal eine Ausnahme, wo die Action mehr braucht. Das kann imho das ODM nicht abfedern - das ODM weiß ja nicht welche Felder in welcher Action rein müssen. Das weiß nur der jeweilige ActionBuilder.

Reply:
Ich verstehe das gesamte Problem leider noch nicht ....

Reply:
Konkretes Beispiel:
Du hast einen CT Warenkorb, der im Property `ShippingInfo` die Versandmethode "DHL" hinterlegt hat. Neben der Versandmethode enthält die `ShippingInfo` aber auch die `TaxCategory`. Nun änderst du die Versandmethode auf "UPS":
```
        $shippingInfo = $cart->getShippingInfo();
        $shippingInfo->setShippingMethodName('UPS');
        $shippingInfo->setShippingRate($rate = new ShippingRate());
        $shippingInfo->setTaxCategory($taxCategory)
        $rate->setIsMatching(true);
        $rate->setPrice($money);
```

Dementsprechend muss das ODM eine Action für die Änderung der Versandmethode bereitstellen. In den meisten fällen ändert sich aber die TaxCategory nicht beim Wechsel der Versandmethode. Diese bleibt also gleich. Daher taucht die TaxCategory auch nicht bei den "$changedData" auf - die TaxCategory wird aber zwingend in der Action benötigt. Daher merge ich unveränderte mit den veränderten Daten, damit ich alle Informationen für die Action haben.